scipy/signal/signaltools.py

```diff
@@ -428,7 +455,177 @@ def fftconvolve(in1, in2, mode="full"):
+        big_O_constant = 10963.92823819 if x.ndim == 1 else 8899.1104874
+    elif mode == 'same':
+        out_shape = x.shape
+        oneD_big_O = {True: 7183.41306773, False: 856.78174111}
+        big_O_constant = oneD_big_O[h.size <= x.size] if x.ndim == 1 \
+                                                      else 34519.21021589
+    elif mode == 'valid':
+        shape_diff = _prod([n - k for n, k in zip(x.shape, h.shape)])
+        out_shape = [n - k + 1 for n, k in zip(x.shape, h.shape)]
+        big_O_constant = 41954.28006344 if x.ndim == 1 else 66453.24316434
+    else:
+        raise ValueError('mode is invalid')
+
+    # see whether the Fourier transform convolution method or the direct
+    # convolution method is faster (discussed in scikit-image PR #1792)
+    direct_time = (x.size * h.size * _prod(out_shape))
+    fft_time = sum_builtin(n * np.log(n) for n in (x.shape + h.shape +
+                                               tuple(out_shape)))
+    return big_O_constant * fft_time < direct_time
+
+
+def _reverse_and_conj(x):
+    """
+    Reverse array `x` in all dimensions and perform the complex conjugate
+    """
+    reverse = [slice(None, None, -1)] * x.ndim
+    return x[reverse].conj()
+
+
+def _np_conv_ok(volume, kernel, mode):
+    """
+    See if numpy supports convolution of x and y (if both are 1D ndarrays and
+    of the appropriate shape).  Numpy's 'same' mode uses the size of the
+    larger input, not the first.
+    """
+    np_conv_ok = volume.ndim == kernel.ndim == 1
+    return np_conv_ok and (volume.size >= kernel.size or mode != 'same')
+
+
+def choose_conv_method(in1, in2, mode='full', method='auto', try_=False):
+    """
+    A method to find the fastest convolution method.
+
+    Parameters
+    ----------
+    in1 : array_like
+        The first argument passed into the convolution function.
+    in2 : array_like
+        The second argument passed into the convolution function.
+    mode : str {'full', 'valid', 'same'}, optional
+        A string indicating the size of the output:
+
+        ``full``
+           The output is the full discrete linear convolution
+           of the inputs. (Default)
+        ``valid``
+           The output consists only of those elements that do not
+           rely on the zero-padding.
+        ``same``
+           The output is the same size as `in1`, centered
+           with respect to the 'full' output.
+    method : str {'fft', 'auto', 'direct'}, optional
+        A str indicating which convolution method would like to be used.
+    try_ : bool, optional
+        If True, run and time the convolution with `in1` and `in2` under both
+        methods and return the fastest method. If False (default), predict the
+        fastest method using precomputed values.
+
+    Returns
+    -------
+    method : str
+        A string indicating which convolution method is fastest, either
+        'direct' or 'fft'
```

Comment:
It should probably also return the time for each. You could say that in `mode='estimate'` the time can only be estimated up to a constant scale factor. Or give it a different name like "cost" and say it's measured in time in `mode='measure'` and operations in `mode='estimate'`.